File: config/node.go
// Package config provides configuration types and loaders for metalog processes.
//
// It supports YAML-based configuration with environment variable overrides.
// The central type is [NodeConfig], which defines database connections, storage
// backends, server settings, and coordinator HA parameters.
package config

import (
	"fmt"
	"os"
	"os/exec"
	"path/filepath"
	"time"

	"gopkg.in/yaml.v3"

	"github.com/y-scope/metalog/telemetry"
)

// NodeConfig is the top-level configuration loaded from node.yaml.
type NodeConfig struct {
	Database    DatabaseSection     `yaml:"database"`
	Telemetry   telemetry.Config    `yaml:"telemetry"`
	Storage     ObjectStorageConfig `yaml:"storage"`
	GRPC        GRPCConfig          `yaml:"grpc"`
	Worker      WorkerConfig        `yaml:"worker"`
	Coordinator CoordinatorConfig   `yaml:"coordinator"`
	Health      HealthConfig        `yaml:"health"`
	Logging     LoggingConfig       `yaml:"logging"`
}

// LoggingConfig holds logging tuning parameters.
type LoggingConfig struct {
	// FailureLogIntervalSeconds controls how often repeated failure warnings
	// are emitted by periodic loops (liveness, planner, retention). The first
	// failure is always logged immediately; subsequent failures repeat at
	// this interval until recovery. Default: 60.
	FailureLogIntervalSeconds int `yaml:"failureLogIntervalSeconds"`
}

// FailureLogInterval returns the failure log interval as a time.Duration.
func (c *LoggingConfig) FailureLogInterval() time.Duration {
	if c.FailureLogIntervalSeconds <= 0 {
		return 60 * time.Second
	}
	return time.Duration(c.FailureLogIntervalSeconds) * time.Second
}

// DatabaseSection holds primary (RW) and optional replica (RO) pool configs.
type DatabaseSection struct {
	Replica *DatabaseConfig `yaml:"replica" envprefix:"DB_REPLICA_"`
	Primary DatabaseConfig  `yaml:"primary" envprefix:"DB_PRIMARY_"`
}

// HAStrategy selects the liveness detection mode.
type HAStrategy string

const (
	HAStrategyHeartbeat HAStrategy = "heartbeat"
	HAStrategyLease     HAStrategy = "lease"
)

// CoordinatorConfig holds coordinator-specific settings (HA, identity).
type CoordinatorConfig struct {
	Name                          string     `yaml:"name" env:"COORDINATOR_NAME"`
	NodeIDEnvVar                  string     `yaml:"nodeIdEnvVar"`
	TableCompression              string     `yaml:"tableCompression"`
	HAStrategy                    HAStrategy `yaml:"haStrategy"`
	ReconciliationIntervalSeconds int        `yaml:"reconciliationIntervalSeconds"`
	HeartbeatIntervalSeconds      int        `yaml:"heartbeatIntervalSeconds"`
	DeadNodeThresholdSeconds      int        `yaml:"deadNodeThresholdSeconds"`
	LeaseTTLSeconds               int        `yaml:"leaseTtlSeconds"`
	LeaseRenewalIntervalSeconds   int        `yaml:"leaseRenewalIntervalSeconds"`
	Enabled                       bool       `yaml:"enabled"`
}

// HealthConfig controls the HTTP health endpoint.
type HealthConfig struct {
	Enabled bool `yaml:"enabled" env:"HEALTH_ENABLED"`
	Port    int  `yaml:"port" env:"HEALTH_PORT"`
}

// GRPCConfig controls the unified gRPC server and per-service toggles.
// The gRPC server starts if any service is enabled. If the section is absent,
// no gRPC server is started.
type GRPCConfig struct {
	BlockingIngestion *bool `yaml:"blockingIngestion"`
	Port              int   `yaml:"port" env:"GRPC_PORT"`
	Ingestion         bool  `yaml:"ingestion"`
	Admin             bool  `yaml:"admin"`
	Query             bool  `yaml:"query"`
	Metadata          bool  `yaml:"metadata"`
}

// IsBlockingIngestion returns whether gRPC ingestion uses blocking submit.
// Defaults to true if not explicitly set.
func (c *GRPCConfig) IsBlockingIngestion() bool {
	if c.BlockingIngestion == nil {
		return true
	}
	return *c.BlockingIngestion
}

// HasAnyService returns true if at least one gRPC service is enabled.
func (c *GRPCConfig) HasAnyService() bool {
	return c.Ingestion || c.Admin || c.Query || c.Metadata
}

// WorkerConfig holds worker settings.
type WorkerConfig struct {
	ClpBinaryPath            string `yaml:"clpBinaryPath"`
	Concurrency              int    `yaml:"concurrency"`
	ClpProcessTimeoutSeconds int    `yaml:"clpProcessTimeoutSeconds"`
}

// ResolveNodeID reads the node ID from the environment variable specified
// in the config. Falls back to os.Hostname() if the env var is not set.
// Returns an error if no node ID can be determined.
func (c *NodeConfig) ResolveNodeID() (string, error) {
	envVar := c.Coordinator.NodeIDEnvVar
	if envVar == "" {
		envVar = "HOSTNAME"
	}
	if v := os.Getenv(envVar); v != "" {
		return v, nil
	}
	h, err := os.Hostname()
	if err != nil {
		return "", fmt.Errorf("resolve node ID: hostname lookup failed: %w", err)
	}
	if h == "" {
		return "", fmt.Errorf("resolve node ID: hostname is empty and $%s is not set", envVar)
	}
	return h, nil
}

// LoadNodeConfig reads and parses a YAML node configuration file.
// Defaults and validation are driven by which sections are present in the YAML:
// coordinator section present → coordinator defaults/validation applied,
// worker section with concurrency > 0 → workers enabled, etc.
func LoadNodeConfig(path string) (*NodeConfig, error) {
	data, err := os.ReadFile(path)
	if err != nil {
		return nil, fmt.Errorf("read config %s: %w", path, err)
	}
	var cfg NodeConfig
	if err := yaml.Unmarshal(data, &cfg); err != nil {
		return nil, fmt.Errorf("parse config %s: %w", path, err)
	}
	if err := ApplyEnvOverrides(&cfg); err != nil {
		return nil, fmt.Errorf("env overrides: %w", err)
	}
	if err := cfg.validateRaw(); err != nil {
		return nil, fmt.Errorf("invalid config: %w", err)
	}
	applyDefaults(&cfg)
	if err := cfg.validate(); err != nil {
		return nil, fmt.Errorf("invalid config: %w", err)
	}
	return &cfg, nil
}

// EffectiveReplica returns the replica database config if set,
// otherwise falls back to the primary database config.
func (c *NodeConfig) EffectiveReplica() DatabaseConfig {
	if c.Database.Replica != nil && c.Database.Replica.Host != "" {
		return *c.Database.Replica
	}
	return c.Database.Primary
}

// HasCoordinator returns true if the coordinator subsystem is enabled.
func (c *NodeConfig) HasCoordinator() bool {
	return c.Coordinator.Enabled
}

// validateRaw checks user-provided values before defaults are applied.
func (c *NodeConfig) validateRaw() error {
	if c.HasCoordinator() {
		if c.Coordinator.LeaseTTLSeconds < 0 {
			return fmt.Errorf("coordinator.leaseTtlSeconds must be non-negative, got %d", c.Coordinator.LeaseTTLSeconds)
		}
		if c.Coordinator.LeaseRenewalIntervalSeconds < 0 {
			return fmt.Errorf("coordinator.leaseRenewalIntervalSeconds must be non-negative, got %d", c.Coordinator.LeaseRenewalIntervalSeconds)
		}
		if c.Coordinator.HeartbeatIntervalSeconds < 0 {
			return fmt.Errorf("coordinator.heartbeatIntervalSeconds must be non-negative, got %d", c.Coordinator.HeartbeatIntervalSeconds)
		}
		if c.Coordinator.DeadNodeThresholdSeconds < 0 {
			return fmt.Errorf("coordinator.deadNodeThresholdSeconds must be non-negative, got %d", c.Coordinator.DeadNodeThresholdSeconds)
		}
		if c.Coordinator.ReconciliationIntervalSeconds < 0 {
			return fmt.Errorf("coordinator.reconciliationIntervalSeconds must be non-negative, got %d", c.Coordinator.ReconciliationIntervalSeconds)
		}
	}
	return nil
}

func (c *NodeConfig) validate() error {
	hasPrimary := c.Database.Primary.Host != ""
	hasReplica := c.Database.Replica != nil && c.Database.Replica.Host != ""
	if !hasPrimary && !hasReplica {
		return fmt.Errorf("at least one of database.primary or database.replica must be configured")
	}
	if hasPrimary {
		if c.Database.Primary.Port < 1 || c.Database.Primary.Port > 65535 {
			return fmt.Errorf("database.primary.port must be 1-65535, got %d", c.Database.Primary.Port)
		}
	}
	if hasReplica {
		if c.Database.Replica.Port < 1 || c.Database.Replica.Port > 65535 {
			return fmt.Errorf("database.replica.port must be 1-65535, got %d", c.Database.Replica.Port)
		}
	}
	if c.Health.Enabled {
		if c.Health.Port < 1 || c.Health.Port > 65535 {
			return fmt.Errorf("health.port must be 1-65535, got %d", c.Health.Port)
		}
	}
	if c.GRPC.HasAnyService() {
		if c.GRPC.Port < 1 || c.GRPC.Port > 65535 {
			return fmt.Errorf("grpc.port must be 1-65535, got %d", c.GRPC.Port)
		}
		// Ingestion and admin require primary database
		if !hasPrimary && (c.GRPC.Ingestion || c.GRPC.Admin) {
			return fmt.Errorf("database.primary is required when grpc.ingestion or grpc.admin is enabled")
		}
		// Ingestion requires coordinator (tables + batching writer)
		if c.GRPC.Ingestion && !c.HasCoordinator() {
			return fmt.Errorf("grpc.ingestion requires coordinator.enabled=true")
		}
	}

	// Coordinator and workers require primary database
	if !hasPrimary && (c.HasCoordinator() || c.Worker.Concurrency > 0) {
		return fmt.Errorf("database.primary is required when coordinator or worker is enabled")
	}

	// Storage validation — defaultBackend must reference a known backend
	if c.Storage.DefaultBackend != "" {
		if _, ok := c.Storage.Backends[c.Storage.DefaultBackend]; !ok {
			return fmt.Errorf("storage.defaultBackend %q not found in storage.backends", c.Storage.DefaultBackend)
		}
	}

	// Workers require clp-s binary
	if c.Worker.Concurrency > 0 && c.Worker.ClpBinaryPath == "" {
		return fmt.Errorf("clp-s binary not found: set worker.clpBinaryPath or add clp-s to $PATH")
	}

	// Coordinator validation
	if c.HasCoordinator() {
		switch c.Coordinator.HAStrategy {
		case HAStrategyHeartbeat, HAStrategyLease:
		default:
			return fmt.Errorf("coordinator.haStrategy must be 'heartbeat' or 'lease', got %q", c.Coordinator.HAStrategy)
		}
		if c.Coordinator.HAStrategy == HAStrategyLease {
			if c.Coordinator.LeaseRenewalIntervalSeconds >= c.Coordinator.LeaseTTLSeconds {
				return fmt.Errorf("coordinator.leaseRenewalIntervalSeconds (%d) must be less than coordinator.leaseTtlSeconds (%d)",
					c.Coordinator.LeaseRenewalIntervalSeconds, c.Coordinator.LeaseTTLSeconds)
			}
		}
	}

	return nil
}

func applyDefaults(cfg *NodeConfig) {
	if cfg.Health.Port == 0 {
		cfg.Health.Port = 8081
	}
	if cfg.GRPC.Port == 0 {
		cfg.GRPC.Port = 9090
	}
	if cfg.Logging.FailureLogIntervalSeconds == 0 {
		cfg.Logging.FailureLogIntervalSeconds = 60
	}

	// Coordinator defaults — only when coordinator section is present
	if cfg.HasCoordinator() {
		if cfg.Coordinator.ReconciliationIntervalSeconds == 0 {
			cfg.Coordinator.ReconciliationIntervalSeconds = 60
		}
		if cfg.Coordinator.HAStrategy == "" {
			cfg.Coordinator.HAStrategy = HAStrategyHeartbeat
		}
		if cfg.Coordinator.HeartbeatIntervalSeconds == 0 {
			cfg.Coordinator.HeartbeatIntervalSeconds = 30
		}
		if cfg.Coordinator.DeadNodeThresholdSeconds == 0 {
			cfg.Coordinator.DeadNodeThresholdSeconds = 180
		}
		if cfg.Coordinator.LeaseTTLSeconds == 0 {
			cfg.Coordinator.LeaseTTLSeconds = 180
		}
		if cfg.Coordinator.LeaseRenewalIntervalSeconds == 0 {
			cfg.Coordinator.LeaseRenewalIntervalSeconds = 30
		}
	}

	// Storage defaults — auto-provision a local filesystem backend when
	// no storage is configured. This lets minimal configs (e.g., metadata-only
	// nodes or integration tests) work without requiring a storage section.
	if cfg.Storage.DefaultBackend == "" {
		cfg.Storage.DefaultBackend = "local"
		if cfg.Storage.Backends == nil {
			cfg.Storage.Backends = make(map[string]StorageBackendConfig)
		}
		if _, ok := cfg.Storage.Backends["local"]; !ok {
			cfg.Storage.Backends["local"] = StorageBackendConfig{
				Type:     "fs",
				BasePath: filepath.Join(os.TempDir(), "clp-storage"),
			}
		}
	}

	// No worker concurrency default — absence means disabled.
	// Users must explicitly set worker.concurrency > 0 to enable workers.

	if cfg.Worker.Concurrency > 0 {
		// Resolve clp-s binary: explicit path > $PATH lookup
		if cfg.Worker.ClpBinaryPath == "" {
			if p, err := exec.LookPath("clp-s"); err == nil {
				cfg.Worker.ClpBinaryPath = p
			}
		}
		if cfg.Worker.ClpProcessTimeoutSeconds == 0 {
			cfg.Worker.ClpProcessTimeoutSeconds = 300
		}
	}
}

File: crates/metalog-config/src/node.rs
use std::path::PathBuf;

use serde::Deserialize;

use crate::{database::DatabaseSection, storage::ObjectStorageConfig};

/// Top-level node configuration loaded from `node.yaml`.
#[derive(Debug, Clone, Deserialize)]
pub struct NodeConfig {
    pub database: DatabaseSection,

    #[serde(default)]
    pub storage: ObjectStorageConfig,

    #[serde(default)]
    pub grpc: GRPCConfig,

    #[serde(default = "default_kafka_driver")]
    pub kafka_driver: String,

    #[serde(default)]
    pub worker: WorkerConfig,

    #[serde(default)]
    pub coordinator: CoordinatorConfig,

    #[serde(default)]
    pub health: HealthConfig,

    #[serde(default)]
    pub logging: LoggingConfig,

    #[serde(default)]
    pub telemetry: TelemetryConfig,
}

fn default_kafka_driver() -> String {
    "rdkafka".into()
}

impl NodeConfig {
    /// Loads a `NodeConfig` from a YAML file.
    pub fn load(path: &str) -> Result<Self, ConfigError> {
        let content = std::fs::read_to_string(path).map_err(|e| ConfigError::Io(path.into(), e))?;
        let config: NodeConfig =
            serde_yaml::from_str(&content).map_err(|e| ConfigError::Parse(path.into(), e))?;
        config.validate()?;
        Ok(config)
    }

    /// Resolves the node ID from environment or hostname.
    pub fn resolve_node_id(&self) -> Result<String, ConfigError> {
        let env_var = if self.coordinator.node_id_env_var.is_empty() {
            "HOSTNAME"
        } else {
            &self.coordinator.node_id_env_var
        };

        if let Ok(val) = std::env::var(env_var) {
            if !val.is_empty() {
                return Ok(val);
            }
        }

        hostname::get()
            .map_err(|e| ConfigError::Validation(format!("resolve hostname: {e}")))?
            .into_string()
            .map_err(|_| ConfigError::Validation("hostname is not valid UTF-8".into()))
    }

    /// Returns the replica config if set, otherwise the primary.
    pub fn effective_replica(&self) -> &crate::DatabaseConfig {
        self.database
            .replica
            .as_ref()
            .unwrap_or(&self.database.primary)
    }

    fn validate(&self) -> Result<(), ConfigError> {
        if self.health.enabled && self.health.port == 0 {
            return Err(ConfigError::Validation(
                "health.port must be non-zero when enabled".into(),
            ));
        }
        Ok(())
    }
}

/// HA strategy for coordinator liveness detection.
#[derive(Debug, Clone, Copy, PartialEq, Eq, Default, Deserialize)]
#[serde(rename_all = "lowercase")]
pub enum HAStrategy {
    #[default]
    Heartbeat,
    Lease,
}

/// Coordinator role configuration.
#[derive(Debug, Clone, Deserialize)]
pub struct CoordinatorConfig {
    #[serde(default)]
    pub enabled: bool,

    #[serde(default)]
    pub name: String,

    #[serde(default)]
    pub node_id_env_var: String,

    #[serde(default)]
    pub table_compression: String,

    #[serde(default)]
    pub ha_strategy: HAStrategy,

    #[serde(default = "default_reconciliation_interval")]
    pub reconciliation_interval_secs: u64,

    #[serde(default = "default_heartbeat_interval")]
    pub heartbeat_interval_secs: u64,

    #[serde(default = "default_dead_node_threshold")]
    pub dead_node_threshold_secs: u64,

    #[serde(default = "default_lease_ttl")]
    pub lease_ttl_secs: u64,

    #[serde(default = "default_lease_renewal")]
    pub lease_renewal_interval_secs: u64,
}

impl Default for CoordinatorConfig {
    fn default() -> Self {
        Self {
            enabled: false,
            name: String::new(),
            node_id_env_var: String::new(),
            table_compression: String::new(),
            ha_strategy: HAStrategy::default(),
            reconciliation_interval_secs: default_reconciliation_interval(),
            heartbeat_interval_secs: default_heartbeat_interval(),
            dead_node_threshold_secs: default_dead_node_threshold(),
            lease_ttl_secs: default_lease_ttl(),
            lease_renewal_interval_secs: default_lease_renewal(),
        }
    }
}

fn default_reconciliation_interval() -> u64 {
    60
}

fn default_heartbeat_interval() -> u64 {
    30
}

fn default_dead_node_threshold() -> u64 {
    180
}

fn default_lease_ttl() -> u64 {
    180
}

fn default_lease_renewal() -> u64 {
    30
}

/// gRPC server configuration.
#[derive(Debug, Clone, Deserialize)]
pub struct GRPCConfig {
    #[serde(default = "default_grpc_port")]
    pub port: u16,

    /// If None or true, ingestion blocks when the channel is full.
    pub blocking_ingestion: Option<bool>,

    #[serde(default)]
    pub ingestion: ServiceConfig,

    #[serde(default)]
    pub admin: ServiceConfig,

    #[serde(default)]
    pub query: ServiceConfig,

    #[serde(default)]
    pub metadata: ServiceConfig,
}

fn default_grpc_port() -> u16 {
    9090
}

impl Default for GRPCConfig {
    fn default() -> Self {
        Self {
            port: default_grpc_port(),
            blocking_ingestion: None,
            ingestion: ServiceConfig::default(),
            admin: ServiceConfig::default(),
            query: ServiceConfig::default(),
            metadata: ServiceConfig::default(),
        }
    }
}

impl GRPCConfig {
    /// Returns true if ingestion should block (default: true).
    pub fn is_blocking_ingestion(&self) -> bool {
        self.blocking_ingestion.unwrap_or(true)
    }

    /// Returns true if any gRPC service is enabled.
    pub fn has_any_service(&self) -> bool {
        self.ingestion.enabled || self.admin.enabled || self.query.enabled || self.metadata.enabled
    }
}

/// Per-service toggle.
#[derive(Debug, Clone, Deserialize, Default)]
pub struct ServiceConfig {
    #[serde(default)]
    pub enabled: bool,
}

/// Worker pool configuration.
#[derive(Debug, Clone, Deserialize)]
pub struct WorkerConfig {
    #[serde(default)]
    pub clp_binary_path: String,

    #[serde(default)]
    pub concurrency: u32,

    #[serde(default = "default_clp_timeout")]
    pub clp_process_timeout_secs: u64,
}

fn default_clp_timeout() -> u64 {
    300
}

impl Default for WorkerConfig {
    fn default() -> Self {
        Self {
            clp_binary_path: String::new(),
            concurrency: 0,
            clp_process_timeout_secs: default_clp_timeout(),
        }
    }
}

impl WorkerConfig {
    /// Resolves the CLP binary path: explicit config, or search $PATH.
    pub fn resolve_clp_binary(&self) -> Option<PathBuf> {
        if !self.clp_binary_path.is_empty() {
            return Some(PathBuf::from(&self.clp_binary_path));
        }
        which::which("clp-s").ok()
    }
}

/// HTTP health probe configuration.
#[derive(Debug, Clone, Deserialize)]
pub struct HealthConfig {
    #[serde(default)]
    pub enabled: bool,

    #[serde(default = "default_health_port")]
    pub port: u16,
}

fn default_health_port() -> u16 {
    8081
}

impl Default for HealthConfig {
    fn default() -> Self {
        Self {
            enabled: false,
            port: default_health_port(),
        }
    }
}

/// Logging tuning.
#[derive(Debug, Clone, Deserialize)]
pub struct LoggingConfig {
    /// Minimum interval between repeated failure log messages (seconds).
    #[serde(default = "default_failure_log_interval")]
    pub failure_log_interval_secs: u64,
}

fn default_failure_log_interval() -> u64 {
    60
}

impl Default for LoggingConfig {
    fn default() -> Self {
        Self {
            failure_log_interval_secs: default_failure_log_interval(),
        }
    }
}

impl LoggingConfig {
    /// Returns the failure log interval as a `Duration`.
    pub fn failure_log_interval(&self) -> std::time::Duration {
        std::time::Duration::from_secs(self.failure_log_interval_secs)
    }
}

/// Telemetry configuration.
#[derive(Debug, Clone, Deserialize)]
pub struct TelemetryConfig {
    #[serde(default)]
    pub enabled: bool,

    #[serde(default = "default_exporter")]
    pub exporter: String,
}

fn default_exporter() -> String {
    "prometheus".into()
}

impl Default for TelemetryConfig {
    fn default() -> Self {
        Self {
            enabled: false,
            exporter: default_exporter(),
        }
    }
}

/// Configuration errors.
#[derive(Debug, thiserror::Error)]
pub enum ConfigError {
    #[error("config file {0}: {1}")]
    Io(String, std::io::Error),

    #[error("config parse {0}: {1}")]
    Parse(String, serde_yaml::Error),

    #[error("config validation: {0}")]
    Validation(String),
}

#[cfg(test)]
mod tests {
    use super::*;

    fn minimal_yaml() -> &'static str {
        r#"
database:
  primary:
    database: metalog_test
"#
    }

    #[test]
    fn parse_minimal() {
        let cfg: NodeConfig = serde_yaml::from_str(minimal_yaml()).unwrap();
        assert_eq!(cfg.database.primary.database, "metalog_test");
        assert_eq!(cfg.database.primary.host, "localhost");
        assert_eq!(cfg.database.primary.port, 3306);
        assert_eq!(cfg.grpc.port, 9090);
        assert_eq!(cfg.health.port, 8081);
        assert!(!cfg.coordinator.enabled);
    }

    #[test]
    fn blocking_ingestion_default_true() {
        let cfg = GRPCConfig::default();
        assert!(cfg.is_blocking_ingestion());
    }

    #[test]
    fn blocking_ingestion_explicit_false() {
        let yaml = "blocking_ingestion: false\nport: 9090";
        let cfg: GRPCConfig = serde_yaml::from_str(yaml).unwrap();
        assert!(!cfg.is_blocking_ingestion());
    }

    #[test]
    fn ha_strategy_default() {
        let cfg = CoordinatorConfig::default();
        assert_eq!(cfg.ha_strategy, HAStrategy::Heartbeat);
    }

    #[test]
    fn ha_strategy_lease() {
        let yaml = "ha_strategy: lease\nenabled: true";
        let cfg: CoordinatorConfig = serde_yaml::from_str(yaml).unwrap();
        assert_eq!(cfg.ha_strategy, HAStrategy::Lease);
    }

    #[test]
    fn effective_replica_fallback() {
        let cfg: NodeConfig = serde_yaml::from_str(minimal_yaml()).unwrap();
        assert_eq!(
            cfg.effective_replica().database,
            cfg.database.primary.database
        );
    }

    #[test]
    fn has_any_service() {
        let mut cfg = GRPCConfig::default();
        assert!(!cfg.has_any_service());
        cfg.ingestion.enabled = true;
        assert!(cfg.has_any_service());
    }

    #[test]
    fn failure_log_interval() {
        let cfg = LoggingConfig::default();
        assert_eq!(
            cfg.failure_log_interval(),
            std::time::Duration::from_secs(60)
        );
    }

    #[test]
    fn coordinator_defaults() {
        let cfg = CoordinatorConfig::default();
        assert_eq!(cfg.reconciliation_interval_secs, 60);
        assert_eq!(cfg.heartbeat_interval_secs, 30);
        assert_eq!(cfg.dead_node_threshold_secs, 180);
        assert_eq!(cfg.lease_ttl_secs, 180);
        assert_eq!(cfg.lease_renewal_interval_secs, 30);
    }
}
